### PR TITLE
XEP-0333: Chat Markers

### PR DESCRIPTION
**Pull request by dommaes**
issue: https://github.com/jsxc/jsxc/issues/745
XEP: https://xmpp.org/extensions/xep-0333.html

### CODE REVIEW

**Comment by sualko** [comment]
> My idea is to implement XEP-0333 as a plugin

Therefore it would be great if you put your plugin inside the `src/plugins/` folder.

**Comment by dommaes** [comment]
I think I'm finally done with this XEP/pull request.
Would you please consider accepting it?
Thanks a lot!

File: src/plugins/chatMarkers/ChatMarkersPlugin.ts [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+   }
+
+   // add "markable" to message according to XEP-0333
+   private preSendMessageStanzaProcessor = (msg: Message, xmlStanza: Strophe.Builder): Promise<any> => {
+      this.supportsChatMarkers(msg.getPeer()).then(() => {
+         if (msg.getType() !== Message.MSGTYPE.GROUPCHAT) {
+            this.addMarkable(xmlStanza, msg);
+         }
+      });
+      return Promise.resolve([msg, xmlStanza]);
```
Comment (sualko) [inline src/plugins/chatMarkers/ChatMarkersPlugin.ts]: Modifications from addMarkable are lost at this point. Return this.supportsChatMarkers and return the array inside the then function.

File: src/plugins/chatMarkers/ChatMarkersPlugin.ts [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+const MARKABLE = 'markable';
+const RECEIVED = 'received';
+const DISPLAYED = 'displayed';
+const ACKNOWLEDGED = 'acknowledged';
+const ID = 'id';
+const FROM = 'from';
+
+export default class ChatMarkersPlugin extends AbstractPlugin {
+   public static readonly PATH_CLOSED_ENVELOPE_GREEN =
+      '../../../images/plugins/chatMarkers/chatmarkers_closed_envelope_green.svg' as String;
```
Comment (sualko) [inline src/plugins/chatMarkers/ChatMarkersPlugin.ts]: You should use the require syntax here.

**Comment by dommaes** [comment]
I implemented your requested changes and squashed the commits into one. Unfortunately I'm not able to resolve the conflict, as Github keeps being unable to merge the changes in src/bootstrap/plugins.ts ...

**Comment by sualko** [comment]
> as Github keeps being unable to merge the changes

I think something went wrong during the squash, because your pr is changing 192 files, including several deletions.

**Comment by dommaes** [comment]
I managed to squash the commits into one.
Again, would you please consider accepting it?
Thank you!

File: src/plugins/chatMarkers/ChatMarkersPlugin.ts [hunk 3]
```diff
@@ -0,0 +1,410 @@
...
+      this.pluginAPI.getConnection().registerHandler(this.onMarkableMessage, null, 'message');
+      this.pluginAPI.getConnection().registerHandler(this.onChatMarkersMessage, null, 'message');
+
+      this.pluginAPI.registerChatWindowInitializedHook((chatWindow: ChatWindow, contact: Contact) => {
+         this.chatWindowInitializedHook(chatWindow, contact);
+      });
+
+      this.data.registerHook('lastReceivedMsgId', (lastReceivedMsgId: string) => {
+         let from = new JID(this.data.get('lastReceivedFrom'));
+         this.supportsChatMarkers(from).then(() => {
```
Comment (sualko) [inline src/plugins/chatMarkers/ChatMarkersPlugin.ts]: The result of `supportsChatMarkers` is ignored.

File: src/plugins/chatMarkers/ChatMarkersPlugin.ts [hunk 4]
```diff
@@ -0,0 +1,410 @@
...
+         if (msg.getType() === Message.MSGTYPE.CHAT) {
+            this.addInIndicators(msg, contact.getChatWindow().getChatWindowMessage(msg));
+            this.registerInHooks(msg, contact.getChatWindow().getChatWindowMessage(msg));
+         }
+         return Promise.resolve([contact, msg, stanza]);
+      });
+   }
+
+   // send "received" on "markable" message according to XEP-0333
+   private onMarkableMessage = (stanza: string) => {
```
Comment (sualko) [inline src/plugins/chatMarkers/ChatMarkersPlugin.ts]: Section 9 says "It is possible for a sender to leak its presence when updating Chat Markers; therefore, a sender SHOULD NOT send Chat Markers to recipients who are not otherwise authorized to view its presence.". How is this ensured?

File: src/plugins/chatMarkers/ChatMarkersPlugin.ts [hunk 5]
```diff
@@ -0,0 +1,410 @@
...
+      this.pluginAPI.addAfterReceiveMessageProcessor(this.afterReceiveMessageProcessor);
+
+      this.pluginAPI.getConnection().registerHandler(this.onMarkableMessage, null, 'message');
+      this.pluginAPI.getConnection().registerHandler(this.onChatMarkersMessage, null, 'message');
+
+      this.pluginAPI.registerChatWindowInitializedHook((chatWindow: ChatWindow, contact: Contact) => {
+         this.chatWindowInitializedHook(chatWindow, contact);
+      });
+
+      this.data.registerHook('lastReceivedMsgId', (lastReceivedMsgId: string) => {
```
Comment (sualko) [inline src/plugins/chatMarkers/ChatMarkersPlugin.ts]: This hook is executed in every tab, which results in multiple messages. Why are you using hooks anyway?

File: src/plugins/chatMarkers/ChatMarkersPlugin.ts [hunk 6]
```diff
@@ -0,0 +1,410 @@
...
+         to: this.data.get('lastAcknowledgedFrom')
+      }).c(ACKNOWLEDGED, {
+         xmlns: Namespace.get(CHATMARKERS),
+         id: lastAcknowledgedMsgId
+      }));
+   }
+
+   private setMessagesChatMarkersAcknowledged(lastAcknowledgedMsgUid: string) {
+      let contact = this.pluginAPI.getContact(new JID(this.data.get('lastAcknowledgedFrom')));
+      if (contact !== undefined) {
```
Comment (sualko) [inline src/plugins/chatMarkers/ChatMarkersPlugin.ts]: This comparison is not needed, because `undefined` is falsy.

File: src/plugins/chatMarkers/ChatMarkersPlugin.ts [hunk 7]
```diff
@@ -0,0 +1,410 @@
...
+                  this.data.set('lastReceivedMsgId', msgId);
+               }
+            }
+         }
+      }
+      return true;
+   }
+
+   // mark messages, depending on the chat marker we get, according to XEP-0333
+   private onChatMarkersMessage = (stanza: string) => {
```
Comment (sualko) [inline src/plugins/chatMarkers/ChatMarkersPlugin.ts]: This function is huge and could be improved.

File: src/plugins/chatMarkers/ChatMarkersPlugin.ts [hunk 8]
```diff
@@ -0,0 +1,410 @@
...
+         }
+      }
+      return true;
+   }
+
+   private chatWindowInitializedHook(chatWindow?: ChatWindow, contact?: Contact) {
+      let transcript = chatWindow.getTranscript();
+      let msg = transcript.getFirstMessage();
+      while (!!msg) {
+         let chatWindowMsg = chatWindow.getChatWindowMessage(msg) as ChatWindowMessage;
```
Comment (sualko) [inline src/plugins/chatMarkers/ChatMarkersPlugin.ts]: Why is the cast needed?

File: src/plugins/chatMarkers/ChatMarkersPlugin.ts [hunk 9]
```diff
@@ -0,0 +1,410 @@
...
+      });
+   }
+
+   private addOutIndicators(msg: IMessage, chatWindowMsg: ChatWindowMessage) {
+      let element = chatWindowMsg.getElement();
+      if (msg.isChatMarkersAcknowledged()) {
+         element.addClass('jsxc-chatmarkers-acknowledged');
+         element.append($(ChatMarkersPlugin.INDICATOR_TEMPLATE({
+            title: Translation.t('chatmarkers-acknowledged'),
+            src: '../../../images/plugins/chatMarkers/chatmarkers_check_mark_green.svg'
```
Comment (sualko) [inline src/plugins/chatMarkers/ChatMarkersPlugin.ts]: Please use the `require` function, because the path can change in some setups.

File: src/plugins/chatMarkers/ChatMarkersPlugin.ts (hunk 8 above)
Comment (dommaes) [reply src/plugins/chatMarkers/ChatMarkersPlugin.ts]: For type safety, as ChatWindow.getChatWindowMessage() doesn't declare a return type. But as this method returns a ChatWindowMessage for sure, it's probably not necessary to cast.

Comment (sualko) [reply src/plugins/chatMarkers/ChatMarkersPlugin.ts]: The better solution would be to change the signature of `ChatWindow.getChatWindowMessage()` :wink:

**Comment by dommaes** [comment]
Next try. Would you please take a look and consider it?
Thanks!

File: src/plugins/chatMarkers/ChatMarkersPlugin.ts [hunk 10]
```diff
@@ -0,0 +1,466 @@
...
+   RECEIVED,
+   DISPLAYED,
+   ACKNOWLEDGED
+}
+
+export default class ChatMarkersPlugin extends AbstractPlugin {
+   private static readonly INDICATOR_TEMPLATE = require('../../../template/plugins/chatMarkers/chatmarkers-indicator.hbs');
+   private static readonly ACKNOWLEDGE_TEMPLATE = require('../../../template/plugins/chatMarkers/chatmarkers-acknowledge.hbs');
+
+   protected pluginAPI = this.pluginAPI as PluginAPI;
```
Comment (sualko) [inline src/plugins/chatMarkers/ChatMarkersPlugin.ts]: Why is this necessary?

File: src/plugins/chatMarkers/ChatMarkersPlugin.ts [hunk 11]
```diff
@@ -0,0 +1,466 @@
...
+      }).c(ACKNOWLEDGED, {
+         xmlns: Namespace.get(CHATMARKERS),
+         id: lastAcknowledgedMsgId
+      }));
+   }
+
+   private setMessagesChatMarkersAcknowledged(lastAcknowledgedMsgUid: string, from: IJID) {
+      this.pluginAPI.Log.debug(`setting previous messages ${ACKNOWLEDGED}. Yaay! =)`);
+      let contact = this.pluginAPI.getContact(from);
+      if (!!contact) {
```
Comment (sualko) [inline src/plugins/chatMarkers/ChatMarkersPlugin.ts]: It's not required to have two exclamation marks here, but it would reduce the indentation of the following code if you return if there is no contact.

File: src/plugins/chatMarkers/ChatMarkersPlugin.ts [hunk 12]
```diff
@@ -0,0 +1,466 @@
...
+   private onPresence = (contact: Contact, newPresence: Presence, oldPresence: Presence) => {
+      this.data.set('presence', newPresence);
+   }
+
+   private preSendMessageStanzaProcessor = (msg: Message, stanza: Strophe.Builder): Promise<any> => {
+      this.pluginAPI.Log.debug('pre send message stanza processor. Yaay! =)');
+      if (msg.getType() === Message.MSGTYPE.CHAT) {
+         this.supportsChatMarkers(msg.getPeer()).then((hasFeature) => {
+            if (hasFeature
+               && (this.getPresence() === Presence.online
```
Comment (sualko) [inline src/plugins/chatMarkers/ChatMarkersPlugin.ts]: Why do you have to check the presence?

File: src/plugins/chatMarkers/ChatMarkersPlugin.ts [hunk 13]
```diff
@@ -0,0 +1,466 @@
...
+               this.addInIndicators(msg, contact.getChatWindow().getChatWindowMessage(msg));
+               this.registerInHooks(msg, contact.getChatWindow().getChatWindowMessage(msg));
+            }
+         });
+      }
+      return Promise.resolve([contact, msg, stanza]);
+   }
+
+   // send "received" on "markable" message according to XEP-0333
+   private onMarkableMessage = (stanza: string) => {
```
Comment (sualko) [inline src/plugins/chatMarkers/ChatMarkersPlugin.ts]: Would be great if you could reduce the indentation.

File: src/plugins/chatMarkers/ChatMarkersPlugin.ts (hunk 4 above)
Comment (sualko) [reply src/plugins/chatMarkers/ChatMarkersPlugin.ts]: >  How is this ensured?

I still see no solution for this.

File: src/plugins/chatMarkers/ChatMarkersPlugin.ts [hunk 14]
```diff
@@ -0,0 +1,466 @@
...
+               this.pluginAPI.Log.debug(`message has "${ACKNOWLEDGED}" element. Yaay! =)`);
+               this.markMessages(element, stanza, ChatMarkerStatus.ACKNOWLEDGED);
+            }
+         }
+      }
+      return true;
+   }
+
+   // mark messages, depending on the chat marker we get, according to XEP-0333
+   private markMessages(element: JQuery<HTMLElement>, stanza: string, status: ChatMarkerStatus) {
```
Comment (sualko) [inline src/plugins/chatMarkers/ChatMarkersPlugin.ts]: This function could be refactored, because it's confusing.

File: src/plugins/chatMarkers/ChatMarkersPlugin.ts [hunk 15]
```diff
@@ -0,0 +1,466 @@
...
+         this.pluginAPI.Log.debug(`message has "${ID}" attribute (${ID}: ${msgId}). Yaay! =)`);
+         let from = $(stanza).attr(FROM);
+         if (!!from) {
+            this.pluginAPI.Log.debug(`message has "${FROM}" attribute (${FROM}: ${from}). Yaay! =)`);
+            if ($(stanza).attr('type') !== Message.MSGTYPE.GROUPCHAT) {
+               let contact = this.pluginAPI.getContact(new JID(from));
+               if (!!contact) {
+                  let transcript = contact.getChatWindow().getTranscript();
+                  let msg = transcript.getMessage(msgId);
+                  while (!!msg) {
```
Comment (sualko) [inline src/plugins/chatMarkers/ChatMarkersPlugin.ts]: Looks similar to `setMessagesChatMarkersAcknowledged`

**Comment by dommaes** [comment]
I think this time I got, what section 9 is referencing to. I did implement your requested changes, as well.
Would you please consider the PR again?

**Comment by sualko** [comment]
Code looks good, but I can't merge it, because currently I have no time to test this pr and we are already in feature freeze for 4.0. Therefore this feature will go into 4.1. Thanks for your work :+1:.

**Comment by sualko** [comment]
I made my changes in https://github.com/jsxc/jsxc/tree/dommaes-xep0333

Can you test that branch? There are still two things missing:

- [ ] get markers from forwarded messages (carbon copy)
- [ ] get markers from mam

**Comment by sualko** [comment]
Markers are now correctly synced and we have a new "transferred" state, like in many other chat clients. I manually merged your changes. Thanks for your contribution.

2020-04-18 11 03 09 localhost 6ed81923c3ec
displayed, received, transferred (top to bottom)
